File: 07.py
import time
from libs import *

from collections import defaultdict


def main():
    print("--- Running Day 7 ---")

    start = time.perf_counter()

    try:
        with open("07.txt", "r") as f:
            data = f.readlines()
    except Exception as e:
        raise ValueError("Unable to read file") from e

    directories = defaultdict(int)
    path = []

    for line in data:
        words = line.strip().split()
        if words[1] == 'cd':
            if words[2] == '..':
                path.pop()
            else:
                path.append(words[2])
        elif words[1] == 'ls':
            continue
        else:
            if words[0] != 'dir':
                size = int(words[0])
                for i in range(len(path)+1):
                    new_path = '/'.join(path[:i]).replace('//', '/')
                    if new_path != '':
                        directories[new_path] += size

    answer_p1 = 0
    for value in directories.values():
        if value <= 100000:
            answer_p1 += value

    print(f"Output Part 1: {answer_p1}")

    used_space = directories['/']
    free_space = 70000000 - used_space

    sorted_directories = sorted(
        directories.items(), key=lambda x: x[1],
    )
    sorted_output = dict(sorted_directories)

    answer_p2 = 0
    for v in sorted_output.values():
        if free_space + v >= 30000000:
            answer_p2 = v
            break

    print(f"Output Part 2: {answer_p2}")

    end = time.perf_counter()
    rtime = (end - start) * 1000  # sec -> ms
    print(f"Took {format_runtime(rtime)}\n")


if __name__ == "__main__":
    main()

File: 07.js
const fs = require('fs');
const { performance } = require('perf_hooks');
const { formatRuntime } = require('./libs');

async function readInput() {
  const input = await fs.promises.readFile('07.txt', 'utf-8');
  return input.trim().split('\n');
}

const smallestDirectory = (free_space, vals) => {
  for (const v of vals) {
    if (free_space + v >= 30000000) {
      return v;
    }
  }
};

async function puzzle() {
  console.log('--- Running Day 7 ---');

  const start = performance.now();

  const input = await readInput();

  const directories = new Map();
  const path = [];

  input.forEach((line) => {
    const words = line.trim().split(' ');
    if (words[1] === 'cd') {
      if (words[2] === '..') {
        path.pop();
      } else {
        path.push(words[2]);
      }
    } else if (words[1] !== 'ls') {
      if (words[0] !== 'dir') {
        const size = +words[0];
        for (let i = 0; i < path.length + 1; i++) {
          const newPath = path.slice(0, i).join('/').replace('//', '/').trim();
          const origSize = directories.get(newPath);
          if (origSize) {
            directories.set(newPath, origSize + size);
          } else {
            directories.set(newPath, size);
          }
        }
      }
    }
  });

  let answer = 0;
  for (const value of directories.values()) {
    if (value <= 100000) {
      answer += value;
    }
  }

  console.log('Part 1: ' + answer);

  const used_space = directories.get('/');
  const free_space = 70000000 - used_space;

  const vals = [...directories.values()];
  vals.sort();

  answer = smallestDirectory(free_space, vals);

  console.log('Part 2: ' + answer);

  const end = performance.now();
  const rtime = end - start;
  console.log('Took:', formatRuntime(rtime));
}

puzzle();
